example/lib/main.dart

```diff
@@ -1,27 +1,35 @@
 import 'package:flutter/material.dart';
 import 'package:flutter_reactive_ble/flutter_reactive_ble.dart';
 import 'package:flutter_reactive_ble_example/src/ble/ble_device_connector.dart';
+import 'package:flutter_reactive_ble_example/src/ble/ble_device_interactor.dart';
 import 'package:flutter_reactive_ble_example/src/ble/ble_scanner.dart';
 import 'package:flutter_reactive_ble_example/src/ble/ble_status_monitor.dart';
 import 'package:flutter_reactive_ble_example/src/ui/ble_status_screen.dart';
 import 'package:flutter_reactive_ble_example/src/ui/device_list.dart';
 import 'package:provider/provider.dart';
 
+import 'src/ble/ble_logger.dart';
+
 const _themeColor = Colors.lightGreen;
 
 void main() {
   WidgetsFlutterBinding.ensureInitialized();
 
+  final _bleLogger = BleLogger();
   final _ble = FlutterReactiveBle();
-  final _scanner = BleScanner(_ble);
+  final _scanner = BleScanner(_ble, _bleLogger.addToLog);
   final _monitor = BleStatusMonitor(_ble);
-  final _connector = BleDeviceConnector(_ble);
+  final _connector = BleDeviceConnector(_ble, _bleLogger.addToLog);
+  final _serviceDiscoverer =
+      BleServiceDiscoverer(_ble.discoverServices, _bleLogger.addToLog);
   runApp(
     MultiProvider(
       providers: [
         Provider.value(value: _scanner),
         Provider.value(value: _monitor),
         Provider.value(value: _connector),
+        Provider.value(value: _serviceDiscoverer),
+        Provider.value(value: _bleLogger),
```

Comment:
If you are deep in refactoring this, you may want to look into the successor, of sorts, to the "provider" package: [rrousselGit/river_pod](https://github.com/rrousselGit/river_pod). The author of "provider" attempts to overcome some of the flaws of its previous creation. I haven't tried it [yet], but am curious.

Reply:
yess I really like rived pod but decided to not go for refactoring at the moment but maybe in a next pr :)